pkg/controller.v2/controller_status.go

```diff
@@ -84,6 +107,100 @@ func (tc *TFJobController) updateStatus(tfjob *tfv1alpha2.TFJob, rtype tfv1alpha
 	return nil
 }
 
+func (tc *TFJobController) updateStatusNew(tfjob *tfv1alpha2.TFJob, rstatus map[string]v1.PodPhase, wreplicas int, preplicas int) error {
+	if preplicas == 0 {
+		tc.updateStatus(tfjob, tfv1alpha2.TFReplicaTypeWorker, wreplicas)
+	} else {
+		status := make(map[string]int)
+
+		initReplicasStatusesNum(status)
+		workerKey := string(tfv1alpha2.TFReplicaTypeWorker) + "-0"
+		switch rstatus[workerKey] {
+		case v1.PodRunning:
+			status[workerRunningReason] = 1
+		case v1.PodSucceeded:
+			status[workerSucceededReason] = 1
+		case v1.PodPending:
+			status[workerPendingReason] = 1
+		case v1.PodFailed:
+			status[workerFailedReason] = 1
+		}
+		for i := 0; i < preplicas; i++ {
+			psKey := string(tfv1alpha2.TFReplicaTypePS) + "-" + strconv.Itoa(i)
+			switch rstatus[psKey] {
+			case v1.PodRunning:
+				status[psRunningReason]++
+			case v1.PodFailed:
+				status[psFailedReason]++
+			case v1.PodPending:
+				status[psPendingReason]++
+			}
+		}
+		if status[psRunningReason] == preplicas && status[workerRunningReason] == 1 {
+			//Running
+			msg := fmt.Sprintf("TFJob %s is running.", tfjob.Name)
+			now := metav1.Now()
+			tfjob.Status.CompletionTime = &now
+			err := tc.updateTFJobConditions(tfjob, tfv1alpha2.TFJobRunning, tfJobRunningReason, msg)
+			if err != nil {
+				loggerForTFJob(tfjob).Infof("Append tfjob condition error: %v", err)
+				return err
+			}
+		}
+		if status[psRunningReason] == preplicas && status[workerSucceededReason] == 1 {
+			//Succeeded
+			msg := fmt.Sprintf("TFJob %s is successfully completed.", tfjob.Name)
+			now := metav1.Now()
+			tfjob.Status.CompletionTime = &now
+			err := tc.updateTFJobConditions(tfjob, tfv1alpha2.TFJobSucceeded, tfJobSucceededReason, msg)
+			if err != nil {
+				loggerForTFJob(tfjob).Infof("Append tfjob condition error: %v", err)
+				return err
+			}
+
+		}
+		if status[psFailedReason] != 0 || status[workerFailedReason] != 0 {
+			//Failed
+			msg := fmt.Sprintf("TFJob %s is failed.", tfjob.Name)
+			now := metav1.Now()
+			tfjob.Status.CompletionTime = &now
+			err := tc.updateTFJobConditions(tfjob, tfv1alpha2.TFJobFailed, tfJobFailedReason, msg)
+			if err != nil {
+				loggerForTFJob(tfjob).Infof("Append tfjob condition error: %v", err)
+				return err
+			}
+
+		}
+		if status[psRunningReason] == preplicas && status[workerPendingReason] == 1 && tfjob.Status.Conditions[len(tfjob.Status.Conditions)-1].Type == tfv1alpha2.TFJobFailed {
```

Comment:
Where do we restart the TFJob? I do not see the logic about it.

Reply:
Restart is controlled by restart policy .